File: src/anomaly-detection/anomaly-detection.ts
import { CloudCostData } from '../types';
import * as zscores from 'zscores';

class AnomalyDetector {
  private data: CloudCostData[] = [];
  private threshold: number;

  constructor(threshold: number) {
    this.threshold = threshold;
  }

  public addData(costData: CloudCostData): void {
    this.data.push(costData);
  }

  public detectAnomalies(): CloudCostData[] {
    const zScores = zscores.calculate(this.data.map(d => d.cost));
    return this.data.filter((_, i) => Math.abs(zScores[i]) > this.threshold);
  }
}

export default AnomalyDetector;

File: anomaly_detection.py

import numpy as np
import pandas as pd

def calculate_anomaly_score(historical_data):
    mean = historical_data.mean()
    std = historical_data.std()
    z_scores = (historical_data - mean) / std
    anomaly_score = z_scores.abs().max()
    return anomaly_score

def detect_anomalies(historical_data, threshold=3):
    anomalies = historical_data[historical_data.abs() > threshold * historical_data.std()]
    return anomalies

def main():
    # Load historical cost data
    data = pd.read_csv('cost_data.csv')

    # Calculate anomaly score for each day
    scores = data['cost'].apply(calculate_anomaly_score, axis=1)

    # Detect anomalies based on a threshold
    anomalies = detect_anomalies(scores)

    # Save detected anomalies to a CSV file
    anomalies.to_csv('anomalies.csv')

    # Print a summary of the detected anomalies
    print(f"Detected {len(anomalies)} anomalies.")

if __name__ == "__main__":
    main()